New vector API + Functional environments + (vector) wrappers

RedTachyon opened the pull request:
Opening this PR to track the progress of the entire project.

This will be merged when the following features are ready, since they interact with each other strongly:
- New vector API (supporting custom vectorization and soon maybe envpool) (mostly done - #93)
- Functional API as a backend for hardware-accelerated envs (mostly done - #25)
- Implementations of (some) hardware-accelerated envs
- Wrappers done in such a way to properly work with individual and vector envs (most likely separately)

RedTachyon commented:
For people involved in this PR (or working on adjacent things) - I just resolved the merge conflicts that github was detecting, but this will need to be manually updated to follow the new plan of firstly merging the changes into the `experimental` directory. I'll try to do this and bring this branch back into a working state soon so that it's not necessary to copy-paste all the new code into a new PR

RedTachyon commented:
The branch is now stable for new contributions. The plan is to keep features related to vector API and functional environments containted to this branch, and then we will merge it to main when the time comes.

younik commented:
@RedTachyon we briefly discuss this in the past, can we call `reset()` in Vec envs at the beginning of `step` if the episode ended in the previous step, instead of calling it immediately?

pseudo-rnd-thoughts commented:
@younik Within my rewrite of the sync vector env, this is how I have implemented the step function if I understand you correctly

younik commented:
> @younik Within my rewrite of the sync vector env (#196), this is how I have implemented the step function if I understand you correctly

Yes, exactly, thank you!

pseudo-rnd-thoughts commented:
@younik I have a longer discussion on the idea in https://github.com/Farama-Foundation/Gymnasium/issues/32#issuecomment-1365493442

pseudo-rnd-thoughts commented:
@RedTachyon With v0.27.1 released and I don't have plans for another release, would you like to merge this PR such that experimental work can happen on main or would you like it to continue on this branch?

pseudo-rnd-thoughts commented on .pre-commit-config.yaml:
```diff
@@ -51,7 +51,7 @@ repos:
     hooks:
       - id: pydocstyle
-        exclude: ^(gymnasium/envs/box2d)|(gymnasium/envs/classic_control)|(gymnasium/envs/mujoco)|(gymnasium/envs/toy_text)|(tests/envs)|(tests/spaces)|(tests/utils)|(tests/vector)|(tests/wrappers)|(docs/)
+        exclude: ^(gymnasium/envs/box2d)|(gymnasium/envs/classic_control)|(gymnasium/envs/mujoco)|(gymnasium/envs/toy_text)|(tests/experimental)|(tests/envs)|(tests/spaces)|(tests/utils)|(tests/vector)|(tests/wrappers)|(docs/)
```
I want all of the new code to have documentation including the testing

RedTachyon replied:
These tests are a copy-paste of `tests/vector` with minimal changes, I narrowed down the exclusion and we can work these two directories at the same time later (or after the new version is made official)

pseudo-rnd-thoughts commented on gymnasium/envs/classic_control/cartpole.py:
```diff
@@ -128,8 +129,9 @@ def __init__(self, render_mode: Optional[str] = None):
-        assert self.action_space.contains(action), err_msg
+        assert self.action_space.contains(
+            action
+        ), f"{action!r} ({type(action)}) invalid"
```
I believe this causes performance issues if I remember correctly.

RedTachyon replied:
The opposite - if you create `err_msg` ahead of time, then it gets created at every step. If the entire string creation is behind `assert`, then if the `assert` doesn't fire, it's basically ignored

pseudo-rnd-thoughts commented on gymnasium/experimental/wrappers/record_episode_statistics.py:
```diff
@@ -0,0 +1,123 @@
+"""Wrapper that tracks the cumulative rewards and episode lengths."""
```
This should be in `experimental/vector/wrappers`

RedTachyon replied:
Done in #322

pseudo-rnd-thoughts commented on gymnasium/experimental/wrappers/record_episode_statistics.py:
```diff
@@ -0,0 +1,123 @@
+        self.length_queue = deque(maxlen=deque_size)
+        self.is_vector_env = True
+
+    def reset(self, **kwargs):
```
Use function args

RedTachyon replied:
Done in 322

pseudo-rnd-thoughts commented on gymnasium/experimental/wrappers/vector_list_info.py:
```diff
@@ -0,0 +1,108 @@
+"""Wrapper that converts the info format for vec envs into the list format."""
```
Should be in `experimental/vector/wrappers`

pseudo-rnd-thoughts reviewed (changes requested):
Other than minor changes, looks good to merge. 
I still wish to rewrite the vector envs from the ground up but this is basic implementation compare with

pseudo-rnd-thoughts commented on gymnasium/__init__.py:
```diff
@@ -27,6 +34,7 @@
     "register",
     "registry",
     "pprint_registry",
+    "make_vec",
```
For my OCD, move to after "make" so in alphabetical order

pseudo-rnd-thoughts commented on tests/experimental/vector/test_vector_make.py:
```diff
@@ -0,0 +1,59 @@
+import pytest
```
Move to `tests/envs/registration`

RedTachyon replied:
Experimental tests mirror the structure of original vector tests, so I'll keep it that way for now, we can do relocations later (it's tests, so low priority)

pseudo-rnd-thoughts reviewed (changes requested):
I don't think we should be modifying the `tests/utils/test_save_video` and `tests/wrappers/*` in this PR

Otherwise, could you address all of my comments then good to merge

RedTachyon commented:
@pseudo-rnd-thoughts I addressed the comments, reverted the autoreset order change as discussed, and resolved merge conflicts, so it should be ready for final review and hopefully merge
